Maclay Teefey (mjt6vj)
Data Project #2 Final

# DS-2002 – Data Project 2 100 points

The goal of this project is to demonstrate (1) an understanding of and (2) competence creating and 
implementing basic data science systems such as pipelines, scripts, data transformations, APIs, databases 
and cloud services. Submit your project in your GitHub Repo or file drop on Collab. 
Data Projects must be done individually.

## ETL Data Processor

You project should demonstrate your understanding of the differing types of data systems (OLTP/OLAP), 
and how data can be extracted from various source systems (structured, semi-structured, unstructured), 
transformed (cleansed, integrated), and then loaded into a destination system that’s optimized for post 
hoc diagnostic analysis.

# Deliverable

## 1. Design a dimensional data mart that represents a simple business process of your choosing.

a. Examples might include retail sales, inventory management, procurement, order 
management, transportation or hospitality bookings, medical appointments, student 
registration and/or attendance.

b. You may select any business process that interests you, but remember that a 
dimensional data mart provides for the post hoc summarization and historic analysis of 
business transactions that reflect the interaction between various entities (e.g., patients 
& doctors, retailers & customers, students & schools/classes, travelers & airlines/hotels).

In [0]:
import os
import json
#import pymongo
import pyspark.pandas as pd  # This uses Koalas that is included in PySpark version 3.2 or newer.
from pyspark.sql.functions import col
from pyspark.sql.types import StructType, StructField, StringType, TimestampType, BinaryType
from pyspark.sql.types import ByteType, ShortType, IntegerType, LongType, FloatType, DecimalType

The three data sets I will be using and their forms:

1. World sample data set from Oracle in SQL

2. Nobel Prizes downloaded and loaded in as a JSON file (local file system) (link: https://api.nobelprize.org/v1/prize.json)

3. Nobel Prize Laureates from the Nobel Prizes API (taken in as JSON)

However, I have found it impossible to deal with JSON files with spark dataframes. Due to Spark not having a json_normalize function and not allowing edits to the spark dataframes without creating a new dataframe I have found it impossible to do the complex data manipulations I did to make the Nobel Prize API Work. Therefore I will take in both Laureates and Prizes as csv files saved from the pandas dataframes from the stage before I joined it with the dim_date dimension in my midterm project.

The relationship being warehoused is between the countries and the nobel prize winners. The key joining point will be the country name and birth country of the laureate. The required date portion will be the Birth Date section.

## 2. Develop an ETL pipeline that extracts, transforms, and loads data into your data mart.

a. Extract data from one or more SQL database tables; hosted locally or in the Cloud.

b. Retrieve a data file, either from a remote or local file system, converting its original 
format (e.g., CSV, JSON) into a SQL database table.

c. Modify the number of columns from each source to the destination.

d. Provide error messages wherever an operation fails (i.e., Try/Except error handlers).

Getting the Countries table from World Data

Functions for Getting Data From and Setting Data Into Databases Taken from Lab 03

In [0]:
# Azure SQL Server Connection Information #####################
jdbc_hostname = "mjt6vj-mysql.mysql.database.azure.com"
jdbc_port = 3306
src_database = "world"

connection_properties = {
  "user" : "mjt6vj",
  "password" : "GB8m0Sjy0J8Q",
  "driver" : "org.mariadb.jdbc.Driver"
}

# MongoDB Atlas Connection Information ########################
atlas_cluster_name = "sandbox"
atlas_database_name = "adventure_works"
atlas_user_name = "m001-student"
atlas_password = "m001-mongodb-basics"

# Data Files (JSON) Information ###############################
dst_database = "nobel_prize_info"

base_dir = "dbfs:/FileStore/default"
database_dir = f"{base_dir}/{dst_database}"

data_dir = f"{base_dir}/prizesSplit"
batch_dir = f"{data_dir}/batch"
stream_dir = f"{data_dir}/stream"
checkpoint_dir = f"{base_dir}/checkpoints/"

output_bronze = f"{database_dir}/fact_nobel/bronze"
output_silver = f"{database_dir}/fact_nobel/silver"
output_gold   = f"{database_dir}/fact_nobel/gold"

# Delete the Streaming Files ################################## 
dbutils.fs.rm(f"{database_dir}/fact_nobel", True)

# Delete the Database Files ###################################
dbutils.fs.rm(database_dir, True)

Out[131]: False

In [0]:
# ######################################################################################################################
# Use this Function to Fetch a DataFrame from the Azure SQL database server.
# ######################################################################################################################
def get_sql_dataframe(host_name, port, db_name, conn_props, sql_query):
    '''Create a JDBC URL to the Azure SQL Database'''
    jdbcUrl = f"jdbc:mysql://{host_name}:{port}/{db_name}"
    
    '''Invoke the spark.read.jdbc() function to query the database, and fill a Pandas DataFrame.'''
    dframe = spark.read.jdbc(url=jdbcUrl, table=sql_query, properties=conn_props)
    
    return dframe


# ######################################################################################################################
# Use this Function to Fetch a DataFrame from the MongoDB Atlas database server Using PyMongo.
# ######################################################################################################################
def get_mongo_dataframe(user_id, pwd, cluster_name, db_name, collection, conditions, projection, sort):
    '''Create a client connection to MongoDB'''
    mongo_uri = f"mongodb+srv://{user_id}:{pwd}@{cluster_name}.zibbf.mongodb.net/{db_name}?retryWrites=true&w=majority"
    
    client = pymongo.MongoClient(mongo_uri)

    '''Query MongoDB, and fill a python list with documents to create a DataFrame'''
    db = client[db_name]
    if conditions and projection and sort:
        dframe = pd.DataFrame(list(db[collection].find(conditions, projection).sort(sort)))
    elif conditions and projection and not sort:
        dframe = pd.DataFrame(list(db[collection].find(conditions, projection)))
    else:
        dframe = pd.DataFrame(list(db[collection].find()))

    client.close()
    
    return dframe

# ######################################################################################################################
# Use this Function to Create New Collections by Uploading JSON file(s) to the MongoDB Atlas server.
# ######################################################################################################################
def set_mongo_collection(user_id, pwd, cluster_name, db_name, src_file_path, json_files):
    '''Create a client connection to MongoDB'''
    mongo_uri = f"mongodb+srv://{user_id}:{pwd}@{cluster_name}.zibbf.mongodb.net/{db_name}?retryWrites=true&w=majority"
    client = pymongo.MongoClient(mongo_uri)
    db = client[db_name]
    
    '''Read in a JSON file, and Use It to Create a New Collection'''
    for file in json_files:
        db.drop_collection(file)
        json_file = os.path.join(src_file_path, json_files[file])
        with open(json_file, 'r') as openfile:
            json_object = json.load(openfile)
            file = db[file]
            result = file.insert_many(json_object)

    client.close()
    
    return result

First we will get the world data, and the date dimension data

In [0]:
df_world = get_sql_dataframe(jdbc_hostname, jdbc_port, src_database, connection_properties, "world.country")
display(df_world)

Code Name Continent Region SurfaceArea IndepYear Population LifeExpectancy GNP GNPOld LocalName GovernmentForm HeadOfState Capital Code2 ABW Aruba North America Caribbean 193.00 null 103000 78.4 828.00 793.00 Aruba Nonmetropolitan Territory of The Netherlands Beatrix 129 AW AFG Afghanistan Asia Southern and Central Asia 652090.00 1919 22720000 45.9 5976.00 null Afganistan/Afqanestan Islamic Emirate Mohammad Omar 1 AF AGO Angola Africa Central Africa 1246700.00 1975 12878000 38.3 6648.00 7984.00 Angola Republic José Eduardo dos Santos 56 AO AIA Anguilla North America Caribbean 96.00 null 8000 76.1 63.20 null Anguilla Dependent Territory of the UK Elisabeth II 62 AI ALB Albania Europe Southern Europe 28748.00 1912 3401200 71.6 3205.00 2500.00 Shqipëria Republic Rexhep Mejdani 34 AL AND Andorra Europe Southern Europe 468.00 1278 78000 83.5 1630.00 null Andorra Parliamentary Coprincipality 55 AD ANT Netherlands Antilles North America Caribbean 800.00 null 217000 74.7 1941.00 null Nederlandse Antillen Nonmetropolitan Territory of The Netherlands Beatrix 33 AN ARE United Arab Emirates Asia Middle East 83600.00 1971 2441000 74.1 37966.00 36846.00 Al-Imarat al-´Arabiya al-Muttahida Emirate Federation Zayid bin Sultan al-Nahayan 65 AE ARG Argentina South America South America 2780400.00 1816 37032000 75.1 340238.00 323310.00 Argentina Federal Republic Fernando de la Rúa 69 AR ARM Armenia Asia Middle East 29800.00 1991 3520000 66.4 1813.00 1627.00 Hajastan Republic Robert Kotšarjan 126 AM ASM American Samoa Oceania Polynesia 199.00 null 68000 75.1 334.00 null Amerika Samoa US Territory George W. Bush 54 AS ATA Antarctica Antarctica Antarctica 13120000.00 null 0 null 0.00 null – Co-administrated null AQ ATF French Southern territories Antarctica Antarctica 7780.00 null 0 null 0.00 null Terres australes françaises Nonmetropolitan Territory of France Jacques Chirac null TF ATG Antigua and Barbuda North America Caribbean 442.00 1981 68000 70.5 612.00 584.00 Antigua and Barbuda Constitutional Monarchy Elisabeth II 63 AG AUS Australia Oceania Australia and New Zealand 7741220.00 1901 18886000 79.8 351182.00 392911.00 Australia Constitutional Monarchy, Federation Elisabeth II 135 AU AUT Austria Europe Western Europe 83859.00 1918 8091800 77.7 211860.00 206025.00 Österreich Federal Republic Thomas Klestil 1523 AT AZE Azerbaijan Asia Middle East 86600.00 1991 7734000 62.9 4127.00 4100.00 Azärbaycan Federal Republic Heydär Äliyev 144 AZ BDI Burundi Africa Eastern Africa 27834.00 1962 6695000 46.2 903.00 982.00 Burundi/Uburundi Republic Pierre Buyoya 552 BI BEL Belgium Europe Western Europe 30518.00 1830 10239000 77.8 249704.00 243948.00 België/Belgique Constitutional Monarchy, Federation Albert II 179 BE BEN Benin Africa Western Africa 112622.00 1960 6097000 50.2 2357.00 2141.00 Bénin Republic Mathieu Kérékou 187 BJ BFA Burkina Faso Africa Western Africa 274000.00 1960 11937000 46.7 2425.00 2201.00 Burkina Faso Republic Blaise Compaoré 549 BF BGD Bangladesh Asia Southern and Central Asia 143998.00 1971 129155000 60.2 32852.00 31966.00 Bangladesh Republic Shahabuddin Ahmad 150 BD BGR Bulgaria Europe Eastern Europe 110994.00 1908 8190900 70.9 12178.00 10169.00 Balgarija Republic Petar Stojanov 539 BG BHR Bahrain Asia Middle East 694.00 1971 617000 73.0 6366.00 6097.00 Al-Bahrayn Monarchy (Emirate) Hamad ibn Isa al-Khalifa 149 BH BHS Bahamas North America Caribbean 13878.00 1973 307000 71.1 3527.00 3347.00 The Bahamas Constitutional Monarchy Elisabeth II 148 BS BIH Bosnia and Herzegovina Europe Southern Europe 51197.00 1992 3972000 71.5 2841.00 null Bosna i Hercegovina Federal Republic Ante Jelavic 201 BA BLR Belarus Europe Eastern Europe 207600.00 1991 10236000 68.0 13714.00 null Belarus Republic Aljaksandr Lukašenka 3520 BY BLZ Belize North America Central America 22696.00 1981 241000 70.9 630.00 616.00 Belize Constitutional Monarchy Elisabeth II 185 BZ BMU Bermuda North America North America 53.00 null 65000 76.9 2328.00 2190.00 Bermuda Dependent

In [0]:
df_date_dim = get_sql_dataframe(jdbc_hostname, jdbc_port, src_database, connection_properties, "world.dim_date")
display(df_date_dim)

date_key full_date date_name date_name_us date_name_eu day_of_week day_name_of_week day_of_month day_of_year weekday_weekend week_of_year month_name month_of_year is_last_day_of_month calendar_quarter calendar_year calendar_year_month calendar_year_qtr fiscal_month_of_year fiscal_quarter fiscal_year fiscal_year_month fiscal_year_qtr 18171130 1817-11-30 1817/11/30 11/30/1817 30/11/1817 1 Sunday 30 334 Weekend 48 November 11 Y 4 1817 1817-11 1817Q4 5 2 1818 1818-05 1818Q2 18171201 1817-12-01 1817/12/01 12/01/1817 01/12/1817 2 Monday 1 335 Weekday 49 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171202 1817-12-02 1817/12/02 12/02/1817 02/12/1817 3 Tuesday 2 336 Weekday 49 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171203 1817-12-03 1817/12/03 12/03/1817 03/12/1817 4 Wednesday 3 337 Weekday 49 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171204 1817-12-04 1817/12/04 12/04/1817 04/12/1817 5 Thursday 4 338 Weekday 49 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171205 1817-12-05 1817/12/05 12/05/1817 05/12/1817 6 Friday 5 339 Weekday 49 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171206 1817-12-06 1817/12/06 12/06/1817 06/12/1817 7 Saturday 6 340 Weekend 49 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171207 1817-12-07 1817/12/07 12/07/1817 07/12/1817 1 Sunday 7 341 Weekend 49 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171208 1817-12-08 1817/12/08 12/08/1817 08/12/1817 2 Monday 8 342 Weekday 50 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171209 1817-12-09 1817/12/09 12/09/1817 09/12/1817 3 Tuesday 9 343 Weekday 50 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171210 1817-12-10 1817/12/10 12/10/1817 10/12/1817 4 Wednesday 10 344 Weekday 50 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171211 1817-12-11 1817/12/11 12/11/1817 11/12/1817 5 Thursday 11 345 Weekday 50 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171212 1817-12-12 1817/12/12 12/12/1817 12/12/1817 6 Friday 12 346 Weekday 50 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171213 1817-12-13 1817/12/13 12/13/1817 13/12/1817 7 Saturday 13 347 Weekend 50 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171214 1817-12-14 1817/12/14 12/14/1817 14/12/1817 1 Sunday 14 348 Weekend 50 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171215 1817-12-15 1817/12/15 12/15/1817 15/12/1817 2 Monday 15 349 Weekday 51 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171216 1817-12-16 1817/12/16 12/16/1817 16/12/1817 3 Tuesday 16 350 Weekday 51 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171217 1817-12-17 1817/12/17 12/17/1817 17/12/1817 4 Wednesday 17 351 Weekday 51 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171218 1817-12-18 1817/12/18 12/18/1817 18/12/1817 5 Thursday 18 352 Weekday 51 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171219 1817-12-19 1817/12/19 12/19/1817 19/12/1817 6 Friday 19 353 Weekday 51 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171220 1817-12-20 1817/12/20 12/20/1817 20/12/1817 7 Saturday 20 354 Weekend 51 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171221 1817-12-21 1817/12/21 12/21/1817 21/12/1817 1 Sunday 21 355 Weekend 51 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171222 1817-12-22 1817/12/22 12/22/1817 22/12/1817 2 Monday 22 356 Weekday 52 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171223 1817-12-23 1817/12/23 12/23/1817 23/12/1817 3 Tuesday 23 357 Weekday 52 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171224 1817-12-24 1817/12/24 12/24/1817 24/12/1817 4 Wednesday 24 358 Weekday 52 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 18171225 1817-12-25 1817/12/25 12/25/1817 25/12/1817 5 Thursday 25 359 Weekday 52 December 12 N 4 1817 1817-12 1817Q4 6 2 1818 1818-06 1818Q2 1817

For the date dimension, I was able to use the provided code from the lab. I did not use it for the midterm, because my mysql and sqlengine would crash trying to load dates from 1817 to 2021, but for the azure mysql server it didnt time out even with taking over 700s

Now to get Laureate data, Country Conversion Data, and Prize Data

In [0]:
address_laureates = f"{base_dir}/laureates.csv"

df_laureates = spark.read.format('csv').options(header='true', inferSchema='true').load(address_laureates)
display(df_laureates)


_c0 LaureateKey firstname surname BirthDate Country CountryCode BirthLocation gender 0 0 Wilhelm Conrad Röntgen 1845-03-27 Prussia (now Germany) DE Lennep (now Remscheid) male 1 1 Hendrik A. Lorentz 1853-07-18 the Netherlands NL Arnhem male 2 2 Pieter Zeeman 1865-05-25 the Netherlands NL Zonnemaire male 3 3 Henri Becquerel 1852-12-15 France FR Paris male 4 4 Pierre Curie 1859-05-15 France FR Paris male 5 5 Marie Curie 1867-11-07 Russian Empire (now Poland) PL Warsaw female 6 6 Lord Rayleigh 1842-11-12 United Kingdom GB Langford Grove, Maldon, Essex male 7 7 Philipp Lenard 1862-06-07 Hungary (now Slovakia) SK Pressburg (now Bratislava) male 8 8 J.J. Thomson 1856-12-18 United Kingdom GB Cheetham Hill male 9 9 Albert A. Michelson 1852-12-19 Prussia (now Poland) PL Strelno (now Strzelno) male 10 10 Gabriel Lippmann 1845-08-16 Luxembourg LU Hollerich male 11 11 Guglielmo Marconi 1874-04-25 Italy IT Bologna male 12 12 Ferdinand Braun 1850-06-06 Hesse-Kassel (now Germany) DE Fulda male 13 13 Johannes Diderik van der Waals 1837-11-23 the Netherlands NL Leiden male 14 14 Wilhelm Wien 1864-01-13 Prussia (now Russia) RU Gaffken (now Parusnoye) male 15 15 Gustaf Dalén 1869-11-30 Sweden SE Stenstorp male 16 16 Heike Kamerlingh Onnes 1853-09-21 the Netherlands NL Groningen male 17 17 Max von Laue 1879-10-09 Germany DE Pfaffendorf male 18 18 William Bragg 1862-07-02 United Kingdom GB Wigton male 19 19 Lawrence Bragg 1890-03-31 Australia AU Adelaide male 20 20 Charles Glover Barkla 1877-06-07 United Kingdom GB Widnes male 21 21 Max Planck 1858-04-23 Schleswig (now Germany) DE Kiel male 22 22 Johannes Stark 1874-04-15 Germany DE Schickenhof male 23 23 Charles Edouard Guillaume 1861-02-15 Switzerland CH Fleurier male 24 24 Albert Einstein 1879-03-14 Germany DE Ulm male 25 25 Niels Bohr 1885-10-07 Denmark DK Copenhagen male 26 26 Robert A. Millikan 1868-03-22 United States US Morrison, IL male 27 27 Manne Siegbahn 1886-12-03 Sweden SE Örebro male 28 28 James Franck 1882-08-26 Germany DE Hamburg male 29 29 Gustav Hertz 1887-07-22 Germany DE Hamburg male 30 30 Jean Baptiste Perrin 1870-09-30 France FR Lille male 31 31 Arthur H. Compton 1892-09-10 United States US Wooster, OH male 32 32 C.T.R. Wilson 1869-02-14 Scotland GB Glencorse male 33 33 Owen Willans Richardson 1879-04-26 United Kingdom GB Dewsbury male 34 34 Louis de Broglie 1892-08-15 France FR Dieppe male 35 35 Sir Chandrasekhara Venkata Raman 1888-11-07 India IN Tiruchirappalli male 36 36 Werner Heisenberg 1901-12-05 Germany DE Würzburg male 37 37 Erwin Schrödinger 1887-08-12 Austria AT Vienna male 38 38 Paul A.M. Dirac 1902-08-08 United Kingdom GB Bristol male 39 39 James Chadwick 1891-10-20 United Kingdom GB Manchester male 40 40 Victor F. Hess 1883-06-24 Austria AT Peggau male 41 41 Carl D. Anderson 1905-09-03 United States US New York, NY male 42 42 Clinton Davisson 1881-10-22 United States US Bloomington, IL male 43 43 George Paget Thomson 1892-05-03 United Kingdom GB Cambridge male 44 44 Enrico Fermi 1901-09-29 Italy IT Rome male 45 45 Ernest Lawrence 1901-08-08 United States US Canton, SD male 46 46 Otto Stern 1888-02-17 Germany (now Poland) PL Sorau (now Zory) male 47 47 Isidor Isaac Rabi 1898-07-29 Austria-Hungary (now Poland) PL Rymanow male 48 48 Wolfgang Pauli 1900-04-25 Austria AT Vienna male 49 49 Percy W. Bridgman 1882-04-21 United States US Cambridge, MA male 50 50 Edward V. Appleton 1892-09-06 United Kingdom GB Bradford male 51 51 Patrick M.S. Blackett 1897-11-18 United Kingdom GB London male 52 52 Hideki Yukawa 1907-01-23 Japan JP Tokyo male 53 53 Cecil Powell 1903-12-05 United Kingdom GB Tonbridge male 54 54 John Cockcroft 1897-05-27 United Kingdom GB Todmorden male 55 55 Ernest T.S. Walton 1903-10-06 Ireland IE Dungarvan male 56 56 Felix Bloch 1905-10-23 Switzerland CH Zurich male 57 57 E. M. Purcell 1912-08-30 United States US Taylorville, IL male 58 58 Frits Zernike 1888-07-16 the Netherlands NL Amsterdam male 59 59 Max Born 1882-12-11 Germany (now Poland) PL Breslau

In [0]:
address_country = f"{base_dir}/country.csv"

df_countries = spark.read.format('csv').options(header='true', inferSchema='true').load(address_country)
display(df_countries)

_c0 id CountryCode2 CountryCode3 CountryName 0 4 AF AFG Afghanistan 1 8 AL ALB Albania 2 12 DZ DZA Algeria 3 20 AD AND Andorra 4 24 AO AGO Angola 5 28 AG ATG Antigua and Barbuda 6 32 AR ARG Argentina 7 51 AM ARM Armenia 8 36 AU AUS Australia 9 40 AT AUT Austria 10 31 AZ AZE Azerbaijan 11 44 BS BHS Bahamas 12 48 BH BHR Bahrain 13 50 BD BGD Bangladesh 14 52 BB BRB Barbados 15 112 BY BLR Belarus 16 56 BE BEL Belgium 17 84 BZ BLZ Belize 18 204 BJ BEN Benin 19 64 BT BTN Bhutan 20 68 BO BOL Bolivia (Plurinational State of) 21 70 BA BIH Bosnia and Herzegovina 22 72 BW BWA Botswana 23 76 BR BRA Brazil 24 96 BN BRN Brunei Darussalam 25 100 BG BGR Bulgaria 26 854 BF BFA Burkina Faso 27 108 BI BDI Burundi 28 132 CV CPV Cabo Verde 29 116 KH KHM Cambodia 30 120 CM CMR Cameroon 31 124 CA CAN Canada 32 140 CF CAF Central African Republic 33 148 TD TCD Chad 34 152 CL CHL Chile 35 156 CN CHN China 36 170 CO COL Colombia 37 174 KM COM Comoros 38 178 CG COG Congo 39 180 CD COD Congo, Democratic Republic of the 40 188 CR CRI Costa Rica 41 384 CI CIV Côte d'Ivoire 42 191 HR HRV Croatia 43 192 CU CUB Cuba 44 196 CY CYP Cyprus 45 203 CZ CZE Czechia 46 208 DK DNK Denmark 47 262 DJ DJI Djibouti 48 212 DM DMA Dominica 49 214 DO DOM Dominican Republic 50 218 EC ECU Ecuador 51 818 EG EGY Egypt 52 222 SV SLV El Salvador 53 226 GQ GNQ Equatorial Guinea 54 232 ER ERI Eritrea 55 233 EE EST Estonia 56 748 SZ SWZ Eswatini 57 231 ET ETH Ethiopia 58 242 FJ FJI Fiji 59 246 FI FIN Finland 60 250 FR FRA France 61 266 GA GAB Gabon 62 270 GM GMB Gambia 63 268 GE GEO Georgia 64 276 DE DEU Germany 65 288 GH GHA Ghana 66 300 GR GRC Greece 67 308 GD GRD Grenada 68 320 GT GTM Guatemala 69 324 GN GIN Guinea 70 624 GW GNB Guinea-Bissau 71 328 GY GUY Guyana 72 332 HT HTI Haiti 73 340 HN HND Honduras 74 348 HU HUN Hungary 75 352 IS ISL Iceland 76 356 IN IND India 77 360 ID IDN Indonesia 78 364 IR IRN Iran (Islamic Republic of) 79 368 IQ IRQ Iraq 80 372 IE IRL Ireland 81 376 IL ISR Israel 82 380 IT ITA Italy 83 388 JM JAM Jamaica 84 392 JP JPN Japan 85 400 JO JOR Jordan 86 398 KZ KAZ Kazakhstan 87 404 KE KEN Kenya 88 296 KI KIR Kiribati 89 408 KP PRK Korea (Democratic People's Republic of) 90 410 KR KOR Korea, Republic of 91 414 KW KWT Kuwait 92 417 KG KGZ Kyrgyzstan 93 418 LA LAO Lao People's Democratic Republic 94 428 LV LVA Latvia 95 422 LB LBN Lebanon 96 426 LS LSO Lesotho 97 430 LR LBR Liberia 98 434 LY LBY Libya 99 438 LI LIE Liechtenstein 100 440 LT LTU Lithuania 101 442 LU LUX Luxembourg 102 450 MG MDG Madagascar 103 454 MW MWI Malawi 104 458 MY MYS Malaysia 105 462 MV MDV Maldives 106 466 ML MLI Mali 107 470 MT MLT Malta 108 584 MH MHL Marshall Islands 109 478 MR MRT Mauritania 110 480 MU MUS Mauritius 111 484 MX MEX Mexico 112 583 FM FSM Micronesia (Federated States of) 113 498 MD MDA Moldova, Republic of 114 492 MC MCO Monaco 115 496 MN MNG Mongolia 116 499 ME MNE Montenegro 117 504 MA MAR Morocco 118 508 MZ MOZ Mozambique 119 104 MM MMR Myanmar 120 516 NA NAM Namibia 121 520 NR NRU Nauru 122 524 NP NPL Nepal 123 528 NL NLD Netherlands 124 554 NZ NZL New Zealand 125 558 NI NIC Nicaragua 126 562 NE NER Niger 127 566 NG NGA Nigeria 128 807 MK MKD North Macedonia 129 578 NO NOR Norway 130 512 OM OMN Oman 131 586 PK PAK Pakistan 132 585 PW PLW Palau 133 591 PA PAN Panama 134 598 PG PNG Papua New Guinea 135 600 PY PRY Paraguay 136 604 PE PER Peru 137 608 PH PHL Philippines 138 616 PL POL Poland 139 620 PT PRT Portugal 140 634 QA QAT Qatar 141 642 RO ROU Romania 142 643 RU RUS Russian Federation 143 646 RW RWA Rwanda 144 659 KN KNA Saint Kitts and Nevis 145 662 LC LCA Saint Lucia 146 670 VC VCT Saint Vincent and the Grenadines 147 882 WS WSM Samoa 148 674 SM SMR San Marino 149 678 ST STP Sao Tome and Principe 150 682 SA SAU Saudi Arabia 151 686 SN SEN Senegal 152 688 RS SRB Serbia 153 690 SC SYC Seychelles 154 694 SL SLE Sierra Leone 155 702 SG SGP Singapore 156 703 SK SVK Slovakia 157 705 SI SVN Slovenia 158 90 SB SLB Solomon Islands 159 706 SO SOM Somalia 160 710 ZA ZAF

For prize data I will use an autoloader, but first I will make dataframes for each fix to the world and laureate dataframes that will be combined together into bronze silver and gold views for combining laureate data, country data, and then selecting important features for analysis.

In [0]:
from pyspark.sql.functions import col

joinType = "inner"
joinCondition = df_date_dim["full_date"] == df_laureates["birthDate"]
df_bronze = df_laureates.join(df_date_dim, joinCondition, joinType)
df_bronze = df_bronze.select(col("LaureateKey"),col("firstname"), col("surname"), col("gender"), col("CountryCode"), col("date_key"), col("calendar_year"))
df_bronze = df_bronze.withColumnRenamed(
  "date_key", "birth_date_key").withColumnRenamed("calendar_year", "birth_year")
display(df_bronze)

LaureateKey firstname surname gender CountryCode birth_date_key birth_year 559 Theodor Mommsen male DE 18171130 1817 457 Frédéric Passy male FR 18220520 1822 460 Randal Cremer male GB 18280318 1828 456 Henry Dunant male CH 18280508 1828 468 Auguste Beernaert male BE 18290726 1829 568 Paul Heyse male DE 18300315 1830 561 Frédéric Mistral male FR 18300908 1830 562 José Echegaray male ES 18320419 1832 560 Bjørnstjerne Bjørnson male NO 18321208 1832 458 Élie Ducommun male CH 18330219 1833 464 Ernesto Teodoro Moneta male IT 18330920 1833 564 Giosuè Carducci male IT 18350727 1835 161 Adolf von Baeyer male DE 18351031 1835 467 Fredrik Bajer male DK 18370421 1837 13 Johannes Diderik van der Waals male NL 18371123 1837 471 Tobias Asser male NL 18380428 1838 558 Sully Prudhomme male FR 18390316 1839 297 Theodor Kocher male CH 18410825 1841 485 Ferdinand Buisson male FR 18411220 1841 6 Lord Rayleigh male GB 18421112 1842 465 Louis Renault male FR 18430521 1843 459 Albert Gobat male CH 18430521 1843 462 Bertha von Suttner female CZ 18430609 1843 292 Camillo Golgi male IT 18430707 1843 291 Robert Koch male DE 18431211 1843 578 Anatole France male FR 18440416 1844 466 Klas Pontus Arnoldson male SE 18441027 1844 473 Elihu Root male US 18450215 1845 0 Wilhelm Conrad Röntgen male DE 18450327 1845 576 Carl Spitteler male CH 18450424 1845 295 Ilya Mechnikov male UA 18450515 1845 294 Alphonse Laveran male FR 18450618 1845 10 Gabriel Lippmann male LU 18450816 1845 566 Rudolf Eucken male DE 18460105 1846 563 Henryk Sienkiewicz male PL 18460505 1846 166 Otto Wallach male RU 18470327 1847 290 Ivan Pavlov male RU 18490914 1849 12 Ferdinand Braun male DE 18500606 1850 301 Charles Richet male FR 18500826 1850 477 Léon Bourgeois male FR 18510521 1851 293 Santiago Ramón y Cajal male ES 18520501 1852 157 Jacobus H. van 't Hoff male NL 18520830 1852 162 Henri Moissan male FR 18520928 1852 160 Sir William Ramsay male GB 18521002 1852 158 Emil Fischer male DE 18521009 1852 469 Paul Henri d'Estournelles de Constant male FR 18521122 1852 3 Henri Becquerel male FR 18521215 1852 9 Albert A. Michelson male PL 18521219 1852 1 Hendrik A. Lorentz male NL 18530718 1853 165 Wilhelm Ostwald male LV 18530902 1853 298 Albrecht Kossel male DE 18530916 1853 16 Heike Kamerlingh Onnes male NL 18530921 1853 296 Paul Ehrlich male PL 18540314 1854 287 Emil von Behring male PL 18540315 1854 474 Henri La Fontaine male BE 18540422 1854 168 Paul Sabatier male FR 18541105 1854 582 George Bernard Shaw male IE 18560726 1856 8 J.J. Thomson male GB 18561218 1856 487 Frank B. Kellogg male US 18561222 1856 476 Woodrow Wilson male US 18561228 1856 311 Julius Wagner-Jauregg male AT 18570307 1857 288 Ronald Ross male IN 18570513 1857 574 Karl Gjellerup male DK 18570602 1857 575 Henrik Pontoppidan male DK 18570724 1857 317 Sir Charles Sherrington male GB 18571127 1857 486 Ludwig Quidde male DE 18580323 1858 21 Max Planck male DE 18580423 1858 313 Christiaan Eijkman male NL 18580811 1858 463 Theodore Roosevelt male US 18581027 1858 567 Selma Lagerlöf female SE 18581120 1858 159 Svante Arrhenius male SE 18590219 1859 4 Pierre Curie male FR 18590515 1859 573 Verner von Heidenstam male SE 18590706 1859 577 Knut Hamsun male NO 18590804 1859 584 Henri Bergson male FR 18591018 1859 163 Eduard Buchner male DE 18600520 1860 309 Willem Einthoven male ID 18600521 1860 489 Jane Addams female US 18600906 1860 478 Hjalmar Branting male SE 18601123 1860 289 Niels Ryberg Finsen male DK 18601215 1860 23 Charles Edouard Guillaume male CH 18610215 1861 571 Rabindranath Tagore male IN 18610507 1861 314 Sir Frederick Hopkins male GB 18610620 1861 480 Fridtjof Nansen male NO 18611010 1861 483 Aristide Briand male FR 18620328 1862 490 Nicholas Murray Butler male US 18620402 1862 299 Allvar Gullstrand male SE 18620605 1862 7 Philipp Lenard male SK 18620607 1862 18 William Bragg male GB 18620702 1862 569 Maurice Maeterlinck male BE 18620829 1862 570 Gerhart Hauptmann male PL 18621115 1862 492 Arthur Henderson male GB 

In [0]:
joinType = "inner"
joinCondition = df_world["Code"] == df_countries["CountryCode3"]
df_silver = df_world.join(df_countries, joinCondition, joinType)
df_silver = df_silver.select(col("CountryCode2"),col("name"), col("Population"), col("GNP"), col("Continent"))
df_silver = df_silver.withColumnRenamed("CountryCode2", "CountryKey2").withColumnRenamed("name", "Country")
display(df_silver)

CountryKey2 Country Population GNP Continent AF Afghanistan 22720000 5976.00 Asia AO Angola 12878000 6648.00 Africa AL Albania 3401200 3205.00 Europe AD Andorra 78000 1630.00 Europe AE United Arab Emirates 2441000 37966.00 Asia AR Argentina 37032000 340238.00 South America AM Armenia 3520000 1813.00 Asia AG Antigua and Barbuda 68000 612.00 North America AU Australia 18886000 351182.00 Oceania AT Austria 8091800 211860.00 Europe AZ Azerbaijan 7734000 4127.00 Asia BI Burundi 6695000 903.00 Africa BE Belgium 10239000 249704.00 Europe BJ Benin 6097000 2357.00 Africa BF Burkina Faso 11937000 2425.00 Africa BD Bangladesh 129155000 32852.00 Asia BG Bulgaria 8190900 12178.00 Europe BH Bahrain 617000 6366.00 Asia BS Bahamas 307000 3527.00 North America BA Bosnia and Herzegovina 3972000 2841.00 Europe BY Belarus 10236000 13714.00 Europe BZ Belize 241000 630.00 North America BO Bolivia 8329000 8571.00 South America BR Brazil 170115000 776739.00 South America BB Barbados 270000 2223.00 North America BN Brunei 328000 11705.00 Asia BT Bhutan 2124000 372.00 Asia BW Botswana 1622000 4834.00 Africa CF Central African Republic 3615000 1054.00 Africa CA Canada 31147000 598862.00 North America CH Switzerland 7160400 264478.00 Europe CL Chile 15211000 72949.00 South America CN China 1277558000 982268.00 Asia CI Côte d’Ivoire 14786000 11345.00 Africa CM Cameroon 15085000 9174.00 Africa CD Congo, The Democratic Republic of the 51654000 6964.00 Africa CG Congo 2943000 2108.00 Africa CO Colombia 42321000 102896.00 South America KM Comoros 578000 4401.00 Africa CV Cape Verde 428000 435.00 Africa CR Costa Rica 4023000 10226.00 North America CU Cuba 11201000 17843.00 North America CY Cyprus 754700 9333.00 Asia CZ Czech Republic 10278100 55017.00 Europe DE Germany 82164700 2133367.00 Europe DJ Djibouti 638000 382.00 Africa DM Dominica 71000 256.00 North America DK Denmark 5330000 174099.00 Europe DO Dominican Republic 8495000 15846.00 North America DZ Algeria 31471000 49982.00 Africa EC Ecuador 12646000 19770.00 South America EG Egypt 68470000 82710.00 Africa ER Eritrea 3850000 650.00 Africa ES Spain 39441700 553233.00 Europe EE Estonia 1439200 5328.00 Europe ET Ethiopia 62565000 6353.00 Africa FI Finland 5171300 121914.00 Europe FJ Fiji Islands 817000 1536.00 Oceania FR France 59225700 1424285.00 Europe FM Micronesia, Federated States of 119000 212.00 Oceania GA Gabon 1226000 5493.00 Africa GB United Kingdom 59623400 1378330.00 Europe GE Georgia 4968000 6064.00 Asia GH Ghana 20212000 7137.00 Africa GN Guinea 7430000 2352.00 Africa GM Gambia 1305000 320.00 Africa GW Guinea-Bissau 1213000 293.00 Africa GQ Equatorial Guinea 453000 283.00 Africa GR Greece 10545700 120724.00 Europe GD Grenada 94000 318.00 North America GT Guatemala 11385000 19008.00 North America GY Guyana 861000 722.00 South America HN Honduras 6485000 5333.00 North America HR Croatia 4473000 20208.00 Europe HT Haiti 8222000 3459.00 North America HU Hungary 10043200 48267.00 Europe ID Indonesia 212107000 84982.00 Asia IN India 1013662000 447114.00 Asia IE Ireland 3775100 75921.00 Europe IR Iran 67702000 195746.00 Asia IQ Iraq 23115000 11500.00 Asia IS Iceland 279000 8255.00 Europe IL Israel 6217000 97477.00 Asia IT Italy 57680000 1161755.00 Europe JM Jamaica 2583000 6871.00 North America JO Jordan 5083000 7526.00 Asia JP Japan 126714000 3787042.00 Asia KZ Kazakstan 16223000 24375.00 Asia KE Kenya 30080000 9217.00 Africa KG Kyrgyzstan 4699000 1626.00 Asia KH Cambodia 11168000 5121.00 Asia KI Kiribati 83000 40.70 Oceania KN Saint Kitts and Nevis 38000 299.00 North America KR South Korea 46844000 320749.00 Asia KW Kuwait 1972000 27037.00 Asia LA Laos 5433000 1292.00 Asia LB Lebanon 3282000 17121.00 Asia LR Liberia 3154000 2012.00 Africa LY Libyan Arab Jamahiriya 5605000 44806.00 Africa LC Saint Lucia 154000 571.00 North America LI Liechtenstein 32300 1119.00 Europe LK Sri Lanka 18827000 15706.00 Asia LS Lesotho 2153000 1061.00 Africa LT Lithuania 3698500 10692.00 Europe LU Luxembourg 435700 163

In [0]:
joinType = "inner"
joinCondition = df_bronze["CountryCode"] == df_silver["CountryKey2"]
df_silver_final = df_bronze.join(df_silver, joinCondition, joinType)
df_silver_final = df_silver_final.select(col("LaureateKey"),col("firstname"), col("surname"), col("gender"), col("CountryCode"), col("birth_date_key"), col("birth_year"),col("Country"), col("Population"), col("GNP"), col("Continent"))
df_silver_final = df_silver_final.withColumnRenamed("CountryCode", "CountryKey").withColumnRenamed("birth_date_key", "birthDateKey").withColumnRenamed("birth_year", "birthYear").withColumnRenamed("LaureateKey", "laureate_key")
display(df_silver_final)

laureate_key firstname surname gender CountryKey birthDateKey birthYear Country Population GNP Continent 559 Theodor Mommsen male DE 18171130 1817 Germany 82164700 2133367.00 Europe 457 Frédéric Passy male FR 18220520 1822 France 59225700 1424285.00 Europe 460 Randal Cremer male GB 18280318 1828 United Kingdom 59623400 1378330.00 Europe 456 Henry Dunant male CH 18280508 1828 Switzerland 7160400 264478.00 Europe 468 Auguste Beernaert male BE 18290726 1829 Belgium 10239000 249704.00 Europe 568 Paul Heyse male DE 18300315 1830 Germany 82164700 2133367.00 Europe 561 Frédéric Mistral male FR 18300908 1830 France 59225700 1424285.00 Europe 562 José Echegaray male ES 18320419 1832 Spain 39441700 553233.00 Europe 560 Bjørnstjerne Bjørnson male NO 18321208 1832 Norway 4478500 145895.00 Europe 458 Élie Ducommun male CH 18330219 1833 Switzerland 7160400 264478.00 Europe 464 Ernesto Teodoro Moneta male IT 18330920 1833 Italy 57680000 1161755.00 Europe 564 Giosuè Carducci male IT 18350727 1835 Italy 57680000 1161755.00 Europe 161 Adolf von Baeyer male DE 18351031 1835 Germany 82164700 2133367.00 Europe 467 Fredrik Bajer male DK 18370421 1837 Denmark 5330000 174099.00 Europe 13 Johannes Diderik van der Waals male NL 18371123 1837 Netherlands 15864000 371362.00 Europe 471 Tobias Asser male NL 18380428 1838 Netherlands 15864000 371362.00 Europe 558 Sully Prudhomme male FR 18390316 1839 France 59225700 1424285.00 Europe 297 Theodor Kocher male CH 18410825 1841 Switzerland 7160400 264478.00 Europe 485 Ferdinand Buisson male FR 18411220 1841 France 59225700 1424285.00 Europe 6 Lord Rayleigh male GB 18421112 1842 United Kingdom 59623400 1378330.00 Europe 465 Louis Renault male FR 18430521 1843 France 59225700 1424285.00 Europe 459 Albert Gobat male CH 18430521 1843 Switzerland 7160400 264478.00 Europe 462 Bertha von Suttner female CZ 18430609 1843 Czech Republic 10278100 55017.00 Europe 292 Camillo Golgi male IT 18430707 1843 Italy 57680000 1161755.00 Europe 291 Robert Koch male DE 18431211 1843 Germany 82164700 2133367.00 Europe 578 Anatole France male FR 18440416 1844 France 59225700 1424285.00 Europe 466 Klas Pontus Arnoldson male SE 18441027 1844 Sweden 8861400 226492.00 Europe 473 Elihu Root male US 18450215 1845 United States 278357000 8510700.00 North America 0 Wilhelm Conrad Röntgen male DE 18450327 1845 Germany 82164700 2133367.00 Europe 576 Carl Spitteler male CH 18450424 1845 Switzerland 7160400 264478.00 Europe 295 Ilya Mechnikov male UA 18450515 1845 Ukraine 50456000 42168.00 Europe 294 Alphonse Laveran male FR 18450618 1845 France 59225700 1424285.00 Europe 10 Gabriel Lippmann male LU 18450816 1845 Luxembourg 435700 16321.00 Europe 566 Rudolf Eucken male DE 18460105 1846 Germany 82164700 2133367.00 Europe 563 Henryk Sienkiewicz male PL 18460505 1846 Poland 38653600 151697.00 Europe 166 Otto Wallach male RU 18470327 1847 Russian Federation 146934000 276608.00 Europe 290 Ivan Pavlov male RU 18490914 1849 Russian Federation 146934000 276608.00 Europe 12 Ferdinand Braun male DE 18500606 1850 Germany 82164700 2133367.00 Europe 301 Charles Richet male FR 18500826 1850 France 59225700 1424285.00 Europe 477 Léon Bourgeois male FR 18510521 1851 France 59225700 1424285.00 Europe 293 Santiago Ramón y Cajal male ES 18520501 1852 Spain 39441700 553233.00 Europe 157 Jacobus H. van 't Hoff male NL 18520830 1852 Netherlands 15864000 371362.00 Europe 162 Henri Moissan male FR 18520928 1852 France 59225700 1424285.00 Europe 160 Sir William Ramsay male GB 18521002 1852 United Kingdom 59623400 1378330.00 Europe 158 Emil Fischer male DE 18521009 1852 Germany 82164700 2133367.00 Europe 469 Paul Henri d'Estournelles de Constant male FR 18521122 1852 France 59225700 1424285.00 Europe 3 Henri Becquerel male FR 18521215 1852 France 59225700 1424285.00 Europe 9 Albert A. Michelson male PL 18521219 1852 Poland 38653600 151697.00 Europe 1 Hendrik A. Lorentz male NL 18530718 1853 Netherlands 15864000 371362.00 Europe 165 Wilhelm Ostwald male LV 18530902 1853 Lat

In [0]:
df_silver_final.write.format("delta").mode("overwrite").saveAsTable("fact_silver")

In [0]:
%sql
SELECT * FROM fact_silver

laureate_key firstname surname gender CountryKey birthDateKey birthYear Country Population GNP Continent 559 Theodor Mommsen male DE 18171130 1817 Germany 82164700 2133367.00 Europe 457 Frédéric Passy male FR 18220520 1822 France 59225700 1424285.00 Europe 460 Randal Cremer male GB 18280318 1828 United Kingdom 59623400 1378330.00 Europe 456 Henry Dunant male CH 18280508 1828 Switzerland 7160400 264478.00 Europe 468 Auguste Beernaert male BE 18290726 1829 Belgium 10239000 249704.00 Europe 568 Paul Heyse male DE 18300315 1830 Germany 82164700 2133367.00 Europe 561 Frédéric Mistral male FR 18300908 1830 France 59225700 1424285.00 Europe 562 José Echegaray male ES 18320419 1832 Spain 39441700 553233.00 Europe 560 Bjørnstjerne Bjørnson male NO 18321208 1832 Norway 4478500 145895.00 Europe 458 Élie Ducommun male CH 18330219 1833 Switzerland 7160400 264478.00 Europe 464 Ernesto Teodoro Moneta male IT 18330920 1833 Italy 57680000 1161755.00 Europe 564 Giosuè Carducci male IT 18350727 1835 Italy 57680000 1161755.00 Europe 161 Adolf von Baeyer male DE 18351031 1835 Germany 82164700 2133367.00 Europe 467 Fredrik Bajer male DK 18370421 1837 Denmark 5330000 174099.00 Europe 13 Johannes Diderik van der Waals male NL 18371123 1837 Netherlands 15864000 371362.00 Europe 471 Tobias Asser male NL 18380428 1838 Netherlands 15864000 371362.00 Europe 558 Sully Prudhomme male FR 18390316 1839 France 59225700 1424285.00 Europe 297 Theodor Kocher male CH 18410825 1841 Switzerland 7160400 264478.00 Europe 485 Ferdinand Buisson male FR 18411220 1841 France 59225700 1424285.00 Europe 6 Lord Rayleigh male GB 18421112 1842 United Kingdom 59623400 1378330.00 Europe 465 Louis Renault male FR 18430521 1843 France 59225700 1424285.00 Europe 459 Albert Gobat male CH 18430521 1843 Switzerland 7160400 264478.00 Europe 462 Bertha von Suttner female CZ 18430609 1843 Czech Republic 10278100 55017.00 Europe 292 Camillo Golgi male IT 18430707 1843 Italy 57680000 1161755.00 Europe 291 Robert Koch male DE 18431211 1843 Germany 82164700 2133367.00 Europe 578 Anatole France male FR 18440416 1844 France 59225700 1424285.00 Europe 466 Klas Pontus Arnoldson male SE 18441027 1844 Sweden 8861400 226492.00 Europe 473 Elihu Root male US 18450215 1845 United States 278357000 8510700.00 North America 0 Wilhelm Conrad Röntgen male DE 18450327 1845 Germany 82164700 2133367.00 Europe 576 Carl Spitteler male CH 18450424 1845 Switzerland 7160400 264478.00 Europe 295 Ilya Mechnikov male UA 18450515 1845 Ukraine 50456000 42168.00 Europe 294 Alphonse Laveran male FR 18450618 1845 France 59225700 1424285.00 Europe 10 Gabriel Lippmann male LU 18450816 1845 Luxembourg 435700 16321.00 Europe 566 Rudolf Eucken male DE 18460105 1846 Germany 82164700 2133367.00 Europe 563 Henryk Sienkiewicz male PL 18460505 1846 Poland 38653600 151697.00 Europe 166 Otto Wallach male RU 18470327 1847 Russian Federation 146934000 276608.00 Europe 290 Ivan Pavlov male RU 18490914 1849 Russian Federation 146934000 276608.00 Europe 12 Ferdinand Braun male DE 18500606 1850 Germany 82164700 2133367.00 Europe 301 Charles Richet male FR 18500826 1850 France 59225700 1424285.00 Europe 477 Léon Bourgeois male FR 18510521 1851 France 59225700 1424285.00 Europe 293 Santiago Ramón y Cajal male ES 18520501 1852 Spain 39441700 553233.00 Europe 157 Jacobus H. van 't Hoff male NL 18520830 1852 Netherlands 15864000 371362.00 Europe 162 Henri Moissan male FR 18520928 1852 France 59225700 1424285.00 Europe 160 Sir William Ramsay male GB 18521002 1852 United Kingdom 59623400 1378330.00 Europe 158 Emil Fischer male DE 18521009 1852 Germany 82164700 2133367.00 Europe 469 Paul Henri d'Estournelles de Constant male FR 18521122 1852 France 59225700 1424285.00 Europe 3 Henri Becquerel male FR 18521215 1852 France 59225700 1424285.00 Europe 9 Albert A. Michelson male PL 18521219 1852 Poland 38653600 151697.00 Europe 1 Hendrik A. Lorentz male NL 18530718 1853 Netherlands 15864000 371362.00 Europe 165 Wilhelm Ostwald male LV 18530902 1853 Lat

Now onto hot-data for the Prizes csv data

In [0]:
def autoload_to_table(data_source, source_format, table_name, checkpoint_directory):
    query = (spark.readStream
                  .format("cloudFiles")
                  .option("cloudFiles.format", source_format)
                  .option("cloudFiles.schemaLocation", checkpoint_directory)
                  .load(data_source)
                  .writeStream
                  .option("checkpointLocation", checkpoint_directory)
                  .option("mergeSchema", "true")
                  .table(table_name))
    return query

In [0]:
query = autoload_to_table(data_source = f"{data_dir}",
                          source_format = "csv",
                          table_name = "target_table",
                          checkpoint_directory = f"{checkpoint_dir}/target_table")


In [0]:
def block_until_stream_is_ready(query, min_batches=2):
    import time
    while len(query.recentProgress) < min_batches:
        time.sleep(5) # Give it a couple of seconds

    print(f"The stream has processed {len(query.recentProgress)} batchs")

block_until_stream_is_ready(query)

The stream has processed 99 batchs


In [0]:
%sql
SELECT * FROM target_table

PrizeKey date category LaureateKey _rescued_data 400 1986-12-10 chemistry 260 {"_c0":"400","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 401 1986-12-10 economics 677 {"_c0":"401","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 402 1986-12-10 literature 640 {"_c0":"402","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 403 1986-12-10 peace 537 {"_c0":"403","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 404 1986-12-10 physics 124 {"_c0":"404","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 405 1986-12-10 physics 125 {"_c0":"405","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 406 1986-12-10 physics 126 {"_c0":"406","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 407 1986-12-10 medicine 428 {"_c0":"407","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 408 1986-12-10 medicine 429 {"_c0":"408","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 409 1985-12-10 chemistry 256 {"_c0":"409","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 410 1985-12-10 chemistry 257 {"_c0":"410","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 411 1985-12-10 economics 676 {"_c0":"411","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 412 1985-12-10 literature 639 {"_c0":"412","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 413 1985-12-10 physics 123 {"_c0":"413","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 414 1985-12-10 medicine 426 {"_c0":"414","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 415 1985-12-10 medicine 427 {"_c0":"415","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 416 1984-12-10 chemistry 255 {"_c0":"416","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 417 1984-12-10 economics 675 {"_c0":"417","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 418 1984-12-10 literature 638 {"_c0":"418","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 419 1984-12-10 peace 535 {"_c0":"419","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 420 1984-12-10 physics 121 {"_c0":"420","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 421 1984-12-10 physics 122 {"_c0":"421","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 422 1984-12-10 medicine 423 {"_c0":"422","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 423 1984-12-10 medicine 424 {"_c0":"423","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 424 1984-12-10 medicine 425 {"_c0":"424","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 425 1983-12-10 chemistry 254 {"_c0":"425","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 426 1983-12-10 economics 674 {"_c0":"426","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 427 1983-12-10 literature 637 {"_c0":"427","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 428 1983-12-10 peace 534 {"_c0":"428","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 429 1983-12-10 physics 119 {"_c0":"429","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 430 1983-12-10 physics 120 {"_c0":"430","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 431 1983-12-10 medicine 422 {"_c0":"431","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 432 1982-12-10 chemistry 253 {"_c0":"432","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 433 1982-12-10 economics 673 {"_c0":"433","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 434 1982-12-10 literature 636 {"_c0":"434","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 435 1982-12-10 peace 532 {"_c0":"435","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 436 1982-12-10 peace 533 {"_c0":"436","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_5.csv"} 437 1982-12-10 physics 118 {"_c

In [0]:
%sql
DESCRIBE TABLE target_table

col_name data_type comment PrizeKey string null date string null category string null LaureateKey string null _rescued_data string null

In [0]:
%sql
CREATE OR REPLACE TEMP VIEW nobel_bronze_view AS
  SELECT category, count(*) total_awards
  FROM target_table
  GROUP BY category;
  
SELECT * FROM nobel_bronze_view

category total_awards chemistry 190 medicine 224 literature 118 peace 108 economics 92 physics 221

In [0]:
%sql
CREATE OR REPLACE TEMP VIEW nobel_silver_view AS
  SELECT *
  FROM target_table
  INNER JOIN fact_silver
  ON target_table.LaureateKey = fact_silver.laureate_key;


In [0]:
%sql
SELECT * FROM nobel_silver_view;

PrizeKey date category LaureateKey _rescued_data laureate_key firstname surname gender CountryKey birthDateKey birthYear Country Population GNP Continent 941 1902-12-10 literature 559 {"_c0":"941","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 559 Theodor Mommsen male DE 18171130 1817 Germany 82164700 2133367.00 Europe 950 1901-12-10 peace 457 {"_c0":"950","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 457 Frédéric Passy male FR 18220520 1822 France 59225700 1424285.00 Europe 936 1903-12-10 peace 460 {"_c0":"936","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 460 Randal Cremer male GB 18280318 1828 United Kingdom 59623400 1378330.00 Europe 949 1901-12-10 peace 456 {"_c0":"949","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 456 Henry Dunant male CH 18280508 1828 Switzerland 7160400 264478.00 Europe 900 1909-12-10 peace 468 {"_c0":"900","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 468 Auguste Beernaert male BE 18290726 1829 Belgium 10239000 249704.00 Europe 895 1910-12-10 literature 568 {"_c0":"895","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_9.csv"} 568 Paul Heyse male DE 18300315 1830 Germany 82164700 2133367.00 Europe 930 1904-12-10 literature 561 {"_c0":"930","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 561 Frédéric Mistral male FR 18300908 1830 France 59225700 1424285.00 Europe 931 1904-12-10 literature 562 {"_c0":"931","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 562 José Echegaray male ES 18320419 1832 Spain 39441700 553233.00 Europe 935 1903-12-10 literature 560 {"_c0":"935","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 560 Bjørnstjerne Bjørnson male NO 18321208 1832 Norway 4478500 145895.00 Europe 942 1902-12-10 peace 458 {"_c0":"942","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 458 Élie Ducommun male CH 18330219 1833 Switzerland 7160400 264478.00 Europe 914 1907-12-10 peace 464 {"_c0":"914","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 464 Ernesto Teodoro Moneta male IT 18330920 1833 Italy 57680000 1161755.00 Europe 919 1906-12-10 literature 564 {"_c0":"919","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 564 Giosuè Carducci male IT 18350727 1835 Italy 57680000 1161755.00 Europe 924 1905-12-10 chemistry 161 {"_c0":"924","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 161 Adolf von Baeyer male DE 18351031 1835 Germany 82164700 2133367.00 Europe 908 1908-12-10 peace 467 {"_c0":"908","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 467 Fredrik Bajer male DK 18370421 1837 Denmark 5330000 174099.00 Europe 896 1910-12-10 physics 13 {"_c0":"896","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_9.csv"} 13 Johannes Diderik van der Waals male NL 18371123 1837 Netherlands 15864000 371362.00 Europe 890 1911-12-10 peace 471 {"_c0":"890","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_9.csv"} 471 Tobias Asser male NL 18380428 1838 Netherlands 15864000 371362.00 Europe 948 1901-12-10 literature 558 {"_c0":"948","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 558 Sully Prudhomme male FR 18390316 1839 France 59225700 1424285.00 Europe 904 1909-12-10 medicine 297 {"_c0":"904","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 297 Theodor Kocher male CH 18410825 1841 Switzerland 7160400 264478.00 Europe 818 1927-12-10 peace 485 {"_c0":"818","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_9.csv"} 485 Ferdinand Buisson male FR 18411220 1841 France 59225700 1424285.00 Europe 932 1904-12-10 physics 6 {"_c0":"932","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 6 Lord Rayleigh male GB 18421112 1842 United Kingdom 59623400 1378330.00 Europe 915 1907-12-10 peace 465 {"_c0":"915","_file_path":"dbfs:/FileStore/default/prizesSplit/prizes_10.csv"} 465 Louis Renault male FR 18430521 1843 France 59225700 1424285.0

In [0]:
%sql
CREATE OR REPLACE TEMP VIEW nobel_gold_view AS
  SELECT country, COUNT(*) AS num_of_awards
  FROM nobel_silver_view
  GROUP BY country
  ORDER BY num_of_awards DESC;


In [0]:
%sql
SELECT * FROM nobel_gold_view

country num_of_awards United States 285 United Kingdom 104 Germany 84 France 59 Sweden 30 Poland 29 Japan 28 Russian Federation 27 Italy 20 Canada 20 Switzerland 19 Austria 19 Netherlands 19 China 12 Norway 12 Denmark 12 Australia 10 Belgium 9 South Africa 9 Hungary 9 India 8 Spain 7 Israel 6 Czech Republic 6 Egypt 6 Finland 5 Ireland 5 Ukraine 5 Argentina 4 Belarus 4 Turkey 3 Lithuania 3 Mexico 3 New Zealand 3 Pakistan 3 Algeria 2 Chile 2 Iran 2 South Korea 2 Guatemala 2 Liberia 2 Saint Lucia 2 Luxembourg 2 Bosnia and Herzegovina 2 Portugal 2 Colombia 2 Yemen 1 Philippines 1 Greece 1 Slovakia 1 Congo, The Democratic Republic of the 1 Madagascar 1 Ghana 1 Peru 1 Croatia 1 Nigeria 1 Bangladesh 1 Morocco 1 Venezuela 1 Iceland 1 Cyprus 1 Indonesia 1 Azerbaijan 1 Trinidad and Tobago 1 Ethiopia 1 Latvia 1 Macedonia 1 Brazil 1 Kenya 1 Slovenia 1 Bulgaria 1 Costa Rica 1

## 3. Author one or more SQL queries (SELECT statements) to demonstrate proper functionality.

a. SELECT data from at least 3 tables (two dimensions; plus the fact table).

b. Perform some type of aggregation (e.g., SUM, COUNT, AVERAGE). This, of course, 
necessitates some form of grouping operation (e.g., GROUP BY <customer.last_name>).

In [0]:
sql_select_statement = """
SELECT `fact_combined`.Country as country_name,
       `world`.Population as population,
       COUNT(*) as count_of_nobel_prizes,
       COUNT(CASE WHEN `laureate`.gender = "male" then 1 ELSE NULL END) as num_to_men,
       COUNT(CASE WHEN `laureate`.gender = "female" then 1 ELSE NULL END) as num_to_women
FROM `data_project_warehouse`.`fact_combined`
INNER JOIN `data_project_warehouse`.`world`
ON  `fact_combined`.`CountryKey` = `world`.`CountryKey`
INNER JOIN `data_project_warehouse`.`laureate`
ON  `fact_combined`.`LaureateKey` = `laureate`.`LaureateKey`
GROUP BY `world`.`country`
ORDER BY count_of_nobel_prizes DESC;
"""
df_result_one = get_dataframe(user_id, pwd, host_name, src_dbname, sql_select_statement)
df_result_one.head()

The 3 Tables used in this select statment are:

1. fact_combined with the country name
2. world with the country population
3. nobel with the gender of the nobel laurette

I also used another table called countries, but it does not show up in the final product because it was only used to convert the 3 letter country intervals in the world sql file to the nobel's 2 letter country intervals